platform/openide.dialogs/src/org/openide/NotifyDescriptor.java

```diff
@@ -1326,4 +1336,110 @@ public void setSelected(boolean selected) {
+            this.total = total;
+        }
+
+        /**
+         * Total number of nested inputs.
+         * @since 7.63
+         */
+        public int getTotalInputs() {
+            return total;
+        }
+
+        /**
+         * Lazy creates nested input of the given ordinal.
+         * @param number input number from interval &lt;1, totalInputs+1&gt;
+         * @return nested selection list, input line, or null
+         * @since 7.63
+         */
+        public NotifyDescriptor getInput(int number) {
+            NotifyDescriptor step = callback.getInput(number);
+            if (step != null) {
+                if (number - 1 < inputs.size()) {
+                    inputs.set(number - 1, step);
+                } else if (number - 1 == inputs.size()) {
+                    inputs.add(step);
+                } else {
+                    return null;
+                }
+                if (number >= total) {
+                    total = number;
+                }
+            }
+            return step;
+        }
+
+        /**
+         * Returns all created nested inputs.
+         * @since 7.63
+         */
+        public NotifyDescriptor[] getInputs() {
+            return inputs.toArray(new NotifyDescriptor[0]);
+        }
+
+        @Override
+        public Object getMessage() {
+            Object msg = super.getMessage();
+            if (msg != null) {
+                return msg;
+            }
+            JPanel panel = new JPanel();
+            panel.setOpaque (false);
+            panel.setLayout(new java.awt.GridBagLayout());
+
+            NotifyDescriptor input;
+            int i = 0;
+            java.awt.GridBagConstraints gridBagConstraints = null;
+            while ((input = getInput(++i)) != null) {
+                gridBagConstraints = new java.awt.GridBagConstraints();
+                gridBagConstraints.gridx = 0;
+                gridBagConstraints.gridy = i - 1;
+                gridBagConstraints.gridwidth = java.awt.GridBagConstraints.RELATIVE;
+                gridBagConstraints.gridheight = java.awt.GridBagConstraints.RELATIVE;
+                gridBagConstraints.anchor = java.awt.GridBagConstraints.FIRST_LINE_START;
+                panel.add((JPanel)input.getMessage(), gridBagConstraints);
+            }
+            if (gridBagConstraints != null) {
+                gridBagConstraints.weighty = 1.0;
+            }
+
+            this.setMessage(panel);
+            return panel;
+        }
+
+        /**
+         * Callback used to lazy create nested inputs.
+         * @since 7.63
+         */
+        public static interface Callback {
+
+            /**
+             * Lazy creates nested input of the given ordinal.
+             * @param number input ordinal from interval &lt;1, totalInputs+1&gt;
+             * @return nested selection list, input line, or null
+             * @since 7.63
+             */
+            public NotifyDescriptor getInput(int number);
```

Comment:
Maybe add the `ComposedInput` instance as a parameter: avoids chicken-and-egg since callback is passed to the constructor of NotifyDescriptor. The handler might change the overall validity (through createNotificationLineSupport) - and maybe adjust the total number of steps.

I am still thinking about the ordinal as the identification of the composedinput state; with multiple possible paths through the composition, the Callback will need to make different comparisons for different paths - it would be better to have fixed/constant IDs for individual constituents. 
I.e. if the `ComposedInput` had a `String stepId`, it would be easier to `switch` or branch on the 'current id' regardless of whether preceding composition members were skipped or not.

Reply:
As for using ordinals as the identification - client can assign and manage stepId <-> step ordinal mapping if necessary. For simple cases current state should be sufficient.